Forms: Migrate form fields to inner blocks JETPACK-262

aaronrobertshaw opened the pull request:
JETPACK-262

Fixes #42358

⚠️ **This PR is now the sole source of truth for both adding new base form field blocks and migrating fields to them** ⚠️

## Proposed changes:

- Add new base blocks for form fields: 
  - Input
  - Label
  - Option
  - Options
- Migrate all form fields to use new base blocks
    - File field is pending and will be updated to inner blocks in due course
- Fixes for various minor forms bugs on trunk e.g. undo loops, minor style conflicts etc
- Change consent and checkbox fields to share option styles not label/input styles for better consistency across fields

## Does this pull request change what data or activity we track or use?

No changes.

## Testing instructions:



**General Notes**
- Please test undo and redo actions throughout all sequences. A number of improvements in this PR specifically enhance stability during these interactions
- As this PR is still under active development and testing, these test instructions may need to evolve over time.

---

1. 🧱 **Field Insertion & Pasting**

Goal: Confirm all form field blocks insert and behave as expected.

- Insert each of the following field types via the block inserter:
  - Text, Name, Email, Number, Website, Phone, Multiline (Textarea)
  - Dropdown
  - Checkbox, Terms Consent
  - Single Choice (Radio), Multiple Choice (Checkbox), 
  - Date Picker
- Paste a field from one form into another.
- Confirm all fields look correct, show relevant block controls, and function as expected.

⸻

2. 🎨 **Styling: Inputs, Labels, and Options**

Goal: Confirm block and global styles apply and override correctly.

- Apply styles via Global Styles (site editor).
- Apply per-block styles using the block inspector.
- Confirm styles render consistently in both the editor and frontend.
- Test toggling individual styles on/off and reapplying.

⸻

3. 🔁 **Style Syncing Between Fields**

Goal: Verify how styles are shared across synced fields of similar types.

- Update a style on one synced field (e.g., font size or text color for an input field).
- Confirm other synced input fields update accordingly.
- Opt a field out of syncing and confirm it no longer updates.
- Insert a new field and confirm it inherits current synced styles.
- Paste a field from another form into a synced form and observe behavior.

⸻

4. 🔀 **Field Transforms**

Goal: Ensure transforming between field types behaves as expected.

- Transform fields (e.g., Name → Email, Text → Phone).
- Confirm that styles and attributes are preserved only when appropriate.
- Confirm that irrelevant attributes (like a "Name" label value when transforming to an Email field) are discarded.

⸻

5. 🧩 **Form Variations & Patterns**

Goal: Ensure all form patterns insert correctly and render well.

Test inserting:

- Contact Form
- RSVP Form
- Registration Form
- Appointment Form
- Feedback Form
- Lead Capture Form
- Salesforce Lead Form

Verify:

- The default structure is valid and editable.
- Labels, placeholders, and field types are appropriate.

⸻

6. ✨ **Style Variants**

Goal: Validate style variations and layout combinations.

Insert:

- Outlined Style Form
- Animated Style Form

For both, test:
  - Base version
  - Version with field customizations
  - Upgrading from trunk form to refactor form
  - Confirm styles work.
  - Test columns and responsive behavior.

⸻

7. 📦 **Field Nesting**

Goal: Confirm fields can be used within layout blocks.

- Nest fields inside:
  - Group blocks
  - Columns
- Confirm style syncing still works across nesting levels.
- Check alignment and spacing behave as expected.

⸻

8. 🪄 **Automatic Form Wrapping**

Goal: Ensure orphaned fields are auto-wrapped in a Form block.

- Insert a field outside any form block.
- Paste one or more fields into a blank page.
- Confirm automatic wrapping behavior.
- Save and verify layout remains valid.

⸻

9. ⚙️ **Required Field Settings**

Goal: Validate required state control across all field types.

- Toggle required setting for fields via field and inner blocks' block toolbars
- Confirm required indicator displays appropriately.
- Add custom required text and check that it renders.
- Disable the required setting and confirm the indicator disappears.

⸻

10. 🧪 **Deprecations & Migrations**

Goal: Confirm legacy forms upgrade smoothly.

- On trunk, create a post containing all field types and apply styling. 
- Switch to the PR branch.

Confirm:

- Legacy form still displays correctly on the frontend
- No block validation errors in editor
- Fields are migrated to inner block structure
- Styling remains intact
- Save the post and confirm the frontend still matches. 

⸻

11. 🌐 **Additional Coverage**

🧑‍🦯 **Keyboard Accessibility**

- Tab through all form fields and verify correct focus behavior.
- Ensure field labels are associated correctly (label-input relationships).

📱 **Mobile Responsiveness**

- Test forms on small screens (phone/tablet viewports).
- Confirm layouts stack properly and remain usable.

↔️ **RTL Language Support**

Switch the site language to RTL (e.g., Hebrew, Arabic).

Check that:
- Layouts flip direction as expected
- Labels and fields align correctly

♻️ **Reusable Blocks**

- Save a form or field group as a reusable block.
- Insert it into another post.
- Confirm editing and layout integrity.

⚙️ **Editor Performance**

- Insert lots of fields (30+).
- Check for lag or editor slowdown during typing and style changes.

🌍 **Multilingual / i18n Support**

- Switch site language.
- Confirm labels, placeholders, and help texts are translatable and update correctly.

aaronrobertshaw committed:
Add base form field blocks

Introduces foundational form field blocks:
- jetpack/input
- jetpack/label
- jetpack/option
- jetpack/options

These blocks will be reused across all field types using inner blocks.

aaronrobertshaw committed:
changelog

aaronrobertshaw committed:
changelog

aaronrobertshaw committed:
Migrate form field blocks to use new base blocks

Updates legacy form field blocks to use shared inner blocks:
- jetpack/input
- jetpack/label
- jetpack/option
- jetpack/options

This commit also updates the registration of form blocks so the new
shared blocks are available in the editor.

aaronrobertshaw committed:
Fix CSS linting issues

andrewserong commented:
Just jotting down a note from testing: overall it's been working quite well for me so far! There's one issue with the Dropdown Field block that I ran into that looked a bit peculiar so just noting it down for now:

1. In the Site Editor, go to edit a page and add a Registration Form pattern (it includes a Dropdown Field)
2. In the Page tab on the right hand sidebar, click the button next to Template and then Swap Template
3. This opens the modal to select another template. In the preview of the current template, the Dropdown Field appears to go into an infinite loop setting itself up (this doesn't appear to happen on `trunk`):

https://github.com/user-attachments/assets/63d4763b-9858-46bb-b0ea-dc4390ae96ce

<details>

<summary>Screenshot of verbose logging from React in the console</summary>

I'm not _quite_ sure what's going on, but there's lots of verbose logging messages from React:

<img width="1277" alt="image" src="https://github.com/user-attachments/assets/f7427def-67d9-445c-b131-94e48735e8fb" />

</details>

Unrelated to this PR, but I find the ability to edit options within the Dropdown Field pretty challenging. That appears to be the case in both this PR and on trunk, though, but could be good to circle back to further down the track.

aaronrobertshaw committed:
Fix issues with style syncing hook:

- Prevents adding bulk changes when syncing blocks to the undo stack
- Correctly determines the parent field block for option blocks to check
  if syncing is enabled
- Correctly find all option blocks regardless of whether they are a
  field's direct child, e.g. consent, or nested within an options block
  e.g. single/multiple choice fields.

aaronrobertshaw commented:
> the Dropdown Field appears to go into an infinite loop setting itself up (this doesn't appear to happen on trunk):

Thanks for the testing @andrewserong :rocket: 

All patterns at this stage will have legacy blocks that need to run through the block deprecations. I'd recently seen a similar issue to what you described when only some blocks had been migrated previously. It's on my list to dig deeper into but got sidetracked today fixing the style syncing for option blocks across the different field types.

aaronrobertshaw committed:
Make select dropdown field's edit component more robust

aaronrobertshaw commented:
> In the Site Editor, go to edit a page and add a Registration Form pattern (it includes a Dropdown Field)

I hit a block validation error as soon as I attempted to add a Registration Form or the pattern. After tweaking the dropdown field's edit component to be a little more robust, I could replicate the issue you noted. 

My suspicion is that this could be related to the `useSyncStyleAttributes` hook. Adding an early return to that prevented the bouncing but the dropdown field never got setup to be the custom select.

aaronrobertshaw committed:
Prevent style syncing when blocks are reused in previews

andrewserong commented:
Another couple of wrinkles that might have already been captured elsewhere (Label block issues):

Hitting "Enter" from the beginning of a label block adds a new line character, where I'd expect it to insert a paragraph block preceding the parent block:

https://github.com/user-attachments/assets/c44a861b-a877-4663-b6d9-5fc88ae4f448

When removing the text from a Label field altogether, if you click out from the block and attempt to click back into it, it's not possible to start typing to add in label text again. You can get back to it via the up arrow by keyboard, but I imagine it's that we're not able to click the empty rich text field. Maybe we can add placeholder text while the Label block is selected or something like that?

https://github.com/user-attachments/assets/bf02cf99-18ca-41cc-9713-b8a19436bc0f

aaronrobertshaw committed:
Ensure label is editable after being emptied and placeholder shows

aaronrobertshaw committed:
First pass at updating form variations

aaronrobertshaw committed:
Quick clean up of form variations

talldan commented on projects/packages/forms/src/blocks/field-checkbox/index.js:
```diff
@@ -0,0 +1,46 @@
+	},
+	providesContext: {
+		'jetpack/field-required': 'required',
+		'jetpack/field-defaultValue': 'defaultValue',
```
Minor - I feel like `'jetpack/field/defaultValue'` might be preferable to reduce the mix of camel/kebab case mixing. I find it reads a little easier too.

talldan commented on projects/packages/forms/src/blocks/field-consent/edit.js:
```diff
@@ -0,0 +1,181 @@
+	useFormWrapper( props );
+
+	const blockProps = useBlockProps( {
+		id: `jetpack-field-consent-${ instanceId }`,
```
Need to check that the `instanceId` still works now that the `withInstanceId` HOC isn't present.

It'd probably be best to switch to `useInstanceId` where possible.

talldan replied:
Another issue might be that block wrappers have their own ids (`id="block-6bb18e8e-1301-449d-880c-bdc272f0ad2d`), and in some cases the code is try set an id on a block wrapper and that isn't working. 🤔

talldan replied:
I've checked `trunk` and from what I can tell these instance ids do nothing because of the issue I mentioned above - the block wrapper has its own instance id that overrides the one the block edit functions try to provide.

The blocks that try to add these custom ids:
- Consent
- Checkbox
- Single / Multiple Choice

Even looking back at the PR that changed the ids to being specified via block props, #38916, I can see no purpose for them. It doesn't seem like they were added to mimic the frontend or for any accessibility reasons (pairing labels with inputs), the ids are added to divs. There are no styles that target the ids, and it'd be hard to do that because of the instance id.

My suggestion is to remove them from the new blocks. I'll make a PR.

edit: issue now resolved via #43055

aaronrobertshaw replied:
Sounds like a plan. Thanks for digging into the history here 🙇

talldan commented on projects/packages/forms/src/blocks/field-consent/edit.js:
```diff
@@ -0,0 +1,181 @@
+		setAttributes,
+	] );
+
+	const onShareFieldAttributesChange = useCallback(
+		value => {
+			setAttributes( { shareFieldAttributes: value } );
+		},
+		[ setAttributes ]
+	);
+
+	const onConsentTypeChange = useCallback(
+		value => {
+			setAttributes( { consentType: value } );
+		},
+		[ setAttributes ]
+	);
```
I'm not sure it's necessary to wrap these with `useCallback`. I think that's only really needed if the functions are being passed to something (component, effect, other callback) that's also memoized.

As it stands though, the `ToggleControl`, `SelectControl` are not memoized, so they will re-render whenever `ConsentFieldEdit` re-renders anyway.

aaronrobertshaw replied:
I'm not tied to this, my linting was warning about unstable or inline functions being passed as props.

talldan commented on projects/packages/forms/src/blocks/field-multiple-choice/deprecated.js:
```diff
@@ -0,0 +1,54 @@
+		},
+	},
+	isEligible( _attributes, innerBlocks ) {
+		if ( innerBlocks.length !== 2 ) {
+			return true;
+		}
```
I don't think this condition is a good idea, because if another inner block type (e.g. a validation message block) is added in the future to the multiple choice block, `isEligible` would return `true` incorrectly, and that'll cause the migration to run when it shouldn't.

This kind of logic problem with `isEligible` has caused some bad bugs in the gutenberg codebase:
- https://github.com/WordPress/gutenberg/pull/63837
- https://github.com/WordPress/gutenberg/pull/65105

Maybe removing this `if` condition and reducing it to the `innerBlocks.some` check below is enough.

aaronrobertshaw replied:
Either way this `isEligible` check would need to be updated upon the addition of any validation inner block. For the time being, it's slightly quicker to check the length first.

talldan replied:
Ok, though there are ways that will make it more future proof:
- Return true if the are only option blocks, false otherwise
- Return false if there's a label and/or an options block present, true otherwise.

Both ways it wouldn't matter if a future new block type is supported.

talldan commented on projects/packages/forms/src/blocks/field-multiple-choice/edit.js:
```diff
@@ -0,0 +1,63 @@
+		select => select( blockEditorStore ).getBlock( clientId ).innerBlocks,
+		[ clientId ]
+	);
+	const options = innerBlocks?.[ 1 ]?.innerBlocks;
```
This could be made more robust, just in case the blocks are reordered for some reason,

talldan commented on projects/packages/forms/src/blocks/field-select/edit.js:
```diff
@@ -1,32 +1,88 @@
+
+const noop = () => undefined;
+
+export default function DropdownFieldEdit( props ) {
```
Nitpick
```suggestion
export default function SelectFieldEdit( props ) {
```

aaronrobertshaw replied:
I'd like to settle on what the field should be called and use that consistently. The original form components named the block `field-select` however gave the block a human readable label of "Dropdown" which was also used in the code too e.g. `JetpackDropdown`. 

Should this be "dropdown" as that's what most users might expect or "select" given it represents an actual select input?

talldan commented on projects/packages/forms/src/blocks/field-single-choice/deprecated.js:
```diff
@@ -0,0 +1,54 @@
+		},
+	},
+	isEligible( _attributes, innerBlocks ) {
+		if ( innerBlocks.length !== 2 ) {
+			return true;
+		}
```
Same as for the other block, maybe this can be removed and it'll make the `isEligible` check more future-proof.

aaronrobertshaw replied:
As noted [here](https://github.com/Automattic/jetpack/pull/42890/files/e4ff0e2216784a2614d119ab755e380c7cc2fbbf#r2035186652), it won't make the check more future proof as the check below would still have to be updated in the event of further valid inner blocks being added

talldan commented on projects/packages/forms/src/blocks/field-textarea/edit.js:
```diff
@@ -0,0 +1,71 @@
+		allowedBlocks: ALLOWED_INNER_BLOCKS,
+		template,
+		templateLock: 'all',
+		type: 'textarea',
```
Is this prop right?

aaronrobertshaw replied:
Nah, I think that's a rogue leftover after trying some different approaches for the inner Input block. I'll get rid of it. Thanks.

talldan commented on projects/packages/forms/src/contact-form/class-contact-form-field.php:
```diff
@@ -141,11 +169,17 @@ public function __construct( $attributes, $content = null, $form = null ) {
 		if ( ! empty( $attributes['options'] ) && is_string( $attributes['options'] ) ) {
 			$attributes['options'] = array_map( 'trim', explode( ',', $attributes['options'] ) );
 
+			// TODO: Work out where these values are set? Are they actually set anywhere?
```
Could it be set by using a shortcode like `[contact-field type="checkbox-multiple" values="1,2,3" /]`?

Perhaps that was an old version of the shortcode.

I see there's also some code here to construct a shortcode with a values attribute like that:
https://github.com/Automattic/jetpack/blob/27ff4357597a0b385570fd0974ded07c27eafe2f/projects/packages/forms/src/contact-form/class-contact-form.php#L879-L884

aaronrobertshaw replied:
Yeah, I came to the same conclusion it was for BC purposes. Thanks for looking into it.

talldan reviewed (commented):
I've run out of time, so this is a partial review.

aaronrobertshaw commented:
:bug:  **Known Issue** :bug: 

Unfortunately, I think I broke the style syncing when adding some fields to a form. Possibly in: https://github.com/Automattic/jetpack/pull/42889/commits/9954d62f65a4173ea5f02a2fbd05709ff1565c64.

It will be worked on over the next couple of days, so ignore bugs there for the moment.

aaronrobertshaw committed:
Pass required block toolbar control through to Option block for checkbox field

aaronrobertshaw committed:
Clean up form field and inner block attributes

aaronrobertshaw committed:
Remove lodash dependency from dropdown field edit component

aaronrobertshaw committed:
Remove remaining lodash uses in new or updated form blocks

aaronrobertshaw committed:
Remove TODO as Dropdown fields didn't use option styles

aaronrobertshaw committed:
Fix default styling for Consent field option label

aaronrobertshaw committed:
Clean up unrequired TODO item

aaronrobertshaw committed:
Clean some more TODOs

aaronrobertshaw committed:
Restore empty string defaults for input, option, and label blocks

aaronrobertshaw committed:
Revert use sync style hook to previous version retaining undo stack fix

aaronrobertshaw commented:
Further to this [comment](https://github.com/Automattic/jetpack/pull/42890#issuecomment-2789459719), I've reverted the style syncing hook to an earlier state via https://github.com/Automattic/jetpack/pull/42890/commits/28e99d37734aa934c255ec42890414dad36371b2. 

This means that not all blocks (e.g. option blocks) are currently sync'd properly. It is still being worked on.

ramonjd commented on projects/packages/forms/src/contact-form/class-contact-form-field.php:
```diff
@@ -740,9 +879,16 @@ public function render_radio_field( $id, $label, $value, $class, $required, $req
 		$field  = "<div class='contact-form__checkbox-wrap'>";
 		$field .= "<input id='" . esc_attr( $id ) . "' type='checkbox' name='" . esc_attr( $id ) . "' value='" . esc_attr__( 'Yes', 'jetpack-forms' ) . "' " . $class . checked( (bool) $value, true, false ) . ' ' . ( $required ? "required aria-required='true'" : '' ) . "/> \n";
-		$field .= "<label for='" . esc_attr( $id ) . "' class='grunion-field-label checkbox" . ( $this->is_error() ? ' form-error' : '' ) . "' style='" . $this->label_styles . "'>";
+		$field .= "<label for='" . esc_attr( $id ) . "' class='" . esc_attr( $label_class ) . "' style='" . esc_attr( $this->_styles ) . esc_attr( $this->option_styles ) . "'>";
```
```suggestion
		$field .= "<label for='" . esc_attr( $id ) . "' class='" . esc_attr( $label_class ) . "' style='" . esc_attr( $this->label_styles ) . esc_attr( $this->option_styles ) . "'>";
```

aaronrobertshaw replied:
Thanks 🤦

aaronrobertshaw committed:
Fix typo

Co-authored-by: Ramon <ramonjd@users.noreply.github.com>

aaronrobertshaw committed:
Fix linting error in editor.scss

talldan committed:
Update form field style syncing (#43004)

* Refactor and update form field syncing.

- Use a parent react context provider as the source of truth instead of an
arbitrary sibling block.
- Perform a deep equality check before syncing to prevent against spurious
updates.

* When updating the block own attributes, mark the change as not persistent so that the user can undo

* Prime the previousOwnAttributes value so that we avoid updating syncedAttributes unnecessarily

* Update inline docs

* Prime the synced attributes when they are initially unset

* Fix syncing of styles across checkbox and consent options

---------

Co-authored-by: Aaron Robertshaw <60436221+aaronrobertshaw@users.noreply.github.com>

aaronrobertshaw committed:
Add transforms for new inner blocks based fields

andrewserong committed:
Attempt to fix warnings about PhanUndeclaredTypeParameter for WP_Block

andrewserong committed:
Fix linting issue in class-contact-form.php

aaronrobertshaw committed:
Forms: Prevent errors when synced attribute context is not available (#43027)

* Forms: Fix error thrown in previews for synced form fields

* Clean up

* Rename the context type stuff

andrewserong committed:
For html_entity_decode, use ENT_COMPAT since we only care about double quotes for the purposes of JSON decoding

andrewserong committed:
Remove is_array and empty check as we now will always have at least one 'class' attribute on the extra_attrs array

aaronrobertshaw committed:
Clean up standalone checkbox inner blocks template

aaronrobertshaw committed:
Clean up dropdown field inner blocks template

aaronrobertshaw committed:
Add example definitions for field block types

aaronrobertshaw committed:
Allow default checked status for standalone option fields

aaronrobertshaw commented:
I've added some transform and example definitions for fields using the inner blocks structure. 

There are still some questions about whether transformed fields should keep the same label, e.g. transforming a Name field to an Email and still having Name as a label doesn't make a lot of sense to me.

Additionally, there seems to be something off with the field block example definitions as they aren't reflecting the example definitions. I'll circle back to that on Monday.

ramonjd commented on projects/packages/forms/src/contact-form/class-contact-form-field.php:
```diff
@@ -301,33 +333,94 @@ public function render() {
+				$class              .= $class ? ' ' . esc_attr( $input_classes ) : esc_attr( $input_classes );
+				$this->field_classes = $input_classes;
+			}
+			if ( ! empty( $option_classes ) ) {
```
Non urgent question:

In my testing I'm seeing that `optionclasses` passed to `Contact_Form_Field` will be added to the field wrapper element. 

```php
// Attributes passed to `new Contact_Form_Field()`
$attributes = array(
    ....some other attrs
    'optionclasses' => 'has-cheese',
);
```

```html
<div style="" class='grunion-field-checkbox-multiple-wrap lalala-wrap has-cheese-wrap grunion-field-wrap'  >
<fieldset ...>
```

Is that intentional?

ramonjd replied:
Actually forget it. Sorry, I didn't see the `-wrap` suffix.

aaronrobertshaw committed:
Fix choice field deprecation edge case

aaronrobertshaw committed:
Fix handling of required and width attributes for Textarea field.

aaronrobertshaw committed:
Fix label block's placeholder fallbacks

ramonjd commented on projects/packages/forms/src/contact-form/class-contact-form-plugin.php:
```diff
@@ -337,124 +391,196 @@ public static function block_attributes_to_shortcode_attributes( $atts, $type )
+
+				// The following handles when option blocks are a direct inner block for a field e.g. singular checkbox field.
+				if ( 'jetpack/option' === $block_name ) {
+					$atts['label']         = $inner_block['attrs']['label'] ?? $inner_block['attrs']['defaultLabel'] ?? '';
```
Can a `jetpack/option` live next to a `jetpack/label`? 

Just wondering if `$atts['label']` set by the  `jetpack/label` block above will ever be overwritten here.

ramonjd replied:
By the way, I checked in the editor and it doesn't look like `jetpack/option` and `jetpack/label` can be innerblock siblings. 

Just asking though in case I'm hallucinating again 😄

talldan commented:
I've made some undo/redo fixes in https://github.com/Automattic/jetpack/pull/43036, which can be merged into this PR if the changes look good.

Some of these issues are also present in trunk, but by solving them it should hopefully make it easier to spot any undo issues that the new style syncing code causes.

talldan committed:
Fix form undo redo issues (#43036)

* Fix undo stack issue with useFormWrapper

* Fix undo stack issue with text area

* Fix undo stack issue with duplicated useFormWrapper

* Fix undo stack issue with consent field block

* Fix undoing after selecting a block variation causing multiple undo steps

* Fix comment typos

Co-authored-by: Aaron Robertshaw <60436221+aaronrobertshaw@users.noreply.github.com>

* Remove `required` property from template

This is now provided to the child block through context

Co-authored-by: Aaron Robertshaw <60436221+aaronrobertshaw@users.noreply.github.com>

---------

Co-authored-by: Aaron Robertshaw <60436221+aaronrobertshaw@users.noreply.github.com>

ramonjd committed:
Forms: Add PHP unit tests for Migrate form fields to inner blocks JETPACK-262  (#43002)

* inital commit - exploratory

* do_shortcode

* Add test_block_attributes_to_shortcode_attributes_with_styles

* add more simple tests

* remove unnecessary changes

* changelog

* LINTO!

* Adding WP namespaces

* test_gutenblock_render_field_checkbox

* Add test for trim

* Use optionsdata to presume that block_attributes_to_shortcode_attributes has taken place. Let's test that as a unit

* Adding classes and styles in attributes.

* block_attributes_to_shortcode_attributes tests

* text field which has input and label

* Phan complains about \DOMNode::getAttribute. Shut it up.

* Try to quiet Phan again by passing a WP_Block instance

* Attempt to fix a couple of the code coverage warnings

* Try to fix namespace again

* Try using the full namespace for the covers annotation instead of the relative name

---------

Co-authored-by: Andrew Serong <14988353+andrewserong@users.noreply.github.com>

aaronrobertshaw committed:
Clean up TODO as the important style for Outlined form style can not be avoided

aaronrobertshaw committed:
Remove extraneous hook dependency

aaronrobertshaw commented on projects/packages/forms/src/blocks/shared/hooks/use-synced-attributes.js:
```diff
@@ -0,0 +1,192 @@
+		// Use lodash `isEqual` to perform a deep comparison of the attributes.
+		// With `!==` this would return a lot of false positives where different blocks
+		// have objects with different references, but the same inner attribute values.
+		// TODO - replace this with not lodash.
```
I had a quick look at alternatives here. 

Core uses `deepEqual` from the `fast-deep-equal` package. Unfortunately, that isn't already a dependency of the Jetpack Forms package. So given there are a lot of uses of Lodash, replacing this will be best done as part of a concerted effort to scrub Lodash out of the Forms package.

This isn't something we need to address to merge this PR

aaronrobertshaw committed:
Remove useFormWrapper audit TODO - all fields use it

aaronrobertshaw committed:
Remove TODOs already captured by issues

talldan committed:
Remove unused ids from consent, checkbox, single choice and multiple choice fields (#43055)

aaronrobertshaw committed:
Opt out of block gap spacing for Options block

This is an enhancement over trunk however the is some work required to
make this actually work, especially given the shortcode rendering on the
frontend. For the time being this support will explicitly be opted out
of. It can be revisited when the benefit justifies the additional
complexity and work.

aaronrobertshaw committed:
Fix saving and display of dropdown field placeholder

aaronrobertshaw committed:
Fix dropdown field placeholder default

This commit also updates the dropdown field input's styles to make it
easier to click when the placeholder has previously been cleared.

aaronrobertshaw committed:
Fix style linting error

ramonjd committed:
Form inner blocks: supplement existing PHP unit tests (#43076)

* Enhance unit tests for Contact_Form_Block by adding support checks

This commit updates the `test_register_child_blocks` method to include an optional parameter for expected block supports and introduces a new test method, `test_child_blocks_block_supports`, to verify the configuration of block supports for various Jetpack blocks. The data provider for the new test method is also added, ensuring comprehensive testing of block support attributes.

* Add unit test for gutenblock_render_field_radio method

This commit introduces a new test method, `test_gutenblock_gutenblock_render_field_radio`, to validate the rendering of the radio field block in the Contact Form Plugin. The test checks that the generated shortcode matches the expected output, ensuring proper functionality of the radio field block with various attributes and inner blocks.

* changelog

* remove duplicate tests. silly me!

* Apply suggestions from code review

Co-authored-by: Aaron Robertshaw <60436221+aaronrobertshaw@users.noreply.github.com>

* syncing server block supports with frontend

* try again

---------

Co-authored-by: Aaron Robertshaw <60436221+aaronrobertshaw@users.noreply.github.com>

aaronrobertshaw committed:
Allow default checked status for standalone option fields (#43077)

talldan committed:
Fix date picker field incorrect classname (#43059)

* Ensure jetpack-field__input classname applies to date picker input instead of jetpack-field__date

* Remove explicit date type in appointment form variation

* Specify the field type for the date field in the block template

* Iron out inconsistency of date type for date picker inputs

* Tied up text fields array

---------

Co-authored-by: Aaron Robertshaw <60436221+aaronrobertshaw@users.noreply.github.com>

aaronrobertshaw committed:
Revert "Remove deprecated covers annotations in line with #42902"

This reverts commit 38791a8459ac68bb654be2d724530e5473f218dd.

Restoring these in case they were in place for coverage requirements.

ramonjd committed:
Forms: move @covers to class level in PHP unit tests (#43081)

* Hoist @cover annotations to class level.

* changelog

aaronrobertshaw commented:
This PR branch now targets `trunk` as its base. 

The original base #42889 was proving difficult to keep up to date with tweaks to the underlying new form blocks without interfering with testing and fixes on this branch.

talldan committed:
Fix field block examples (#43096)

* Fix field block examples being altered by useFormWrapper

useFormWrapper adds a form wrapper around any field that doesn't have one.

It wasn't preserving the field's own inner blocks, which became visible in the block
examples, where single or multiple choice fields have inner blocks but no wrapper.

This PR ensures the inner blocks are preserved by calling getBlocks to get the inner
blocks of the field and passing that as the inner blocks in the replaceBlocks call
within useFormWrapper.

* Small improvement to variable name

* Add missing useEffect deps

* Fix select/dropdown example

* Switch e2e test close welcome guide code to use preferences

aaronrobertshaw committed:
Forms: Fix display of label and option fallback values in transform previews JETPACK-278 (#43100)

talldan committed:
Add fallback styles for dropdown block when css vars are not populated (#43101)

aaronrobertshaw committed:
Add missed cleaned up label fallback handling

enejb commented on projects/packages/forms/src/blocks/field-telephone/index.js:
```diff
@@ -0,0 +1,48 @@
+import { Icon } from '@wordpress/components';
```
Sometimes I in the jetpack code base I see 
```
import { Icon } from '@wordpress/components';
import { mobile } from '@wordpress/icons';
```
and other times
I see 
```
import { Icon, mobile } from '@wordpress/icons'; 
```
I am assuming the end result is the same thing. Does it make sense to just use the @wordpress/icons package ?

Since that Icon component is a bit more light wight?

aaronrobertshaw replied:
Both Icon components are light weight, so I don't think that aspect makes much difference in the scheme of things.

For me, I'd lean towards using `Icon` from the `@wordpress/components` package:
- This package is effectively a source of truth for WordPress components
- Contains some size logic to help ensure icons align with WordPress design systems
- There's also some validation before attempting to pass through additional props
- More flexible for future updates
- Jetpack already has `@wordpress/components` as a dependency

aaronrobertshaw committed:
Fix rendering of legend as a label

- Ensure block class is added to legend element
- Add custom block classes and styles to legend
- Tweak default style specificity so global styles can apply

aaronrobertshaw committed:
Remove redundant array check to make linter happy.

andrewserong commented on projects/packages/forms/src/blocks/field-telephone/deprecated.js:
```diff
@@ -0,0 +1,11 @@
+		...INNER_BLOCKS_DEPRECATION,
+		attributes: {
+			...INNER_BLOCKS_DEPRECATION.attributes,
+			label: { type: 'string', default: 'Phone' },
```
(Nothing to action here): Tiny drive-by comment while I walk through the changes in the codebase: this `default: 'Phone'` does _not_ need a translate call because the original block did not translate this default label. Just thought I'd mention it in case anyone else runs into it.

andrewserong commented on projects/packages/forms/src/blocks/shared/components/jetpack-field-controls.js:
```diff
@@ -0,0 +1,154 @@
+	const optionColorLabel =
+		blockStyle === 'button'
+			? __( 'Button Text', 'jetpack-forms' )
+			: __( 'Option Text', 'jetpack-forms', 0 );
+
+	const inputColorLabel = isChoicesBlock
+		? optionColorLabel
+		: __( 'Field Text', 'jetpack-forms', 0 );
+
+	const backgroundColorLabel = isChoicesBlock
+		? __( 'Background', 'jetpack-forms' )
+		: __( 'Field Background', 'jetpack-forms', 0 );
```
Are the zeros passed in to `__()` necessary? I thought that function only took two params 🤔

(This is just a drive-by comment, and it isn't urgent. I can fix these up tomorrow or next week if they're not there for a reason)

aaronrobertshaw replied:
They're just copied and pasted from the original `JetpackFieldControls` component. So if it was "doing it wrong" so is this.

I suspect that now the fields have been migrated, those using this component from the `shared` folder won't need all that settings config. It probably isn't even used at the moment 🤷

aaronrobertshaw replied:
I'll push a commit soon to clean up some of the dead code in this component and remove those odd extra params.

andrewserong replied:
> They're just copied and pasted from the original JetpackFieldControls component. So if it was "doing it wrong" so is this [...] I'll push a commit soon to clean up some of the dead code

Ah, that makes sense. Thanks for tidying up! 🍻

aaronrobertshaw replied:
Quick clean up done in https://github.com/Automattic/jetpack/pull/42890/commits/ef2cbe36cacb7b5f335acf8def9a3d3ad761742f. After an equally quick before and after comparison I didn't spot any regressions for field controls.

anomiex replied:
FYI: The 0 is a workaround for a minifier optimization that can break i18n.

GlotPress relies on seeing calls like `__( 'string literal', 'textdomain' )` in order to extract strings for translation. However, the minifier may convert code like
```js
	const backgroundColorLabel = isChoicesBlock
		? __( 'Background', 'jetpack-forms' )
		: __( 'Field Background', 'jetpack-forms' );
```
into
```js
	const backgroundColorLabel = __( isChoicesBlock ? 'Background' : 'Field Background', 'jetpack-forms' );
```
which won't be found by GlotPress because the parameter is no longer a string literal.

Adding the dummy "0" parameter is one way to avoid that transformation. Another would be to use `_x()`, like
```js
	const backgroundColorLabel = isChoicesBlock
		? _x( 'Background', 'Some context text', 'jetpack-forms' )
		: _x( 'Field Background', 'Some different context text', 'jetpack-forms' );
```
(note the two contexts need to be different), or even just
```js
	const backgroundColorLabel = isChoicesBlock
		? __( 'Background', 'jetpack-forms' )
		: _x( 'Field Background', '', 'jetpack-forms' );
```
(note passing an empty string for the context seems to be equivalent to `__()`, but the minifier doesn't know that).

Since this is an easy thing to accidentally have happen, we have [a custom webpack plugin](https://www.npmjs.com/package/@automattic/i18n-check-webpack-plugin) that tries to detect when strings go missing in the production build. 🙂

andrewserong replied:
Thanks for the context @anomiex! In this case, the code could be removed because it wasn't being used, but that's great to know for next time 👍

aaronrobertshaw committed:
Clean up dead code in shared JetpackFieldControls

andrewserong committed:
Try to fix metadata in doc comments warning

andrewserong committed:
Try updating PHP test attributes again, to hopefully please the linter

ramonjd committed:
Forms: ensure custom CSS in global styles applies correctly to checkbox and radio wrappers (#43184)

* Refactor class attribute handling in Contact_Form_Field

- Trimmed type variable to ensure consistent class generation.
- Added specific classes for 'checkbox-multiple' and 'radio' types.
- Included TODOs for potential improvements and testing.

* changelog

* Add block styles for checkbox-multiple and radio fields

- Registered new block styles for 'jetpack/field-checkbox-multiple' and 'jetpack/field-radio'.
- Updated class handling in Contact_Form_Field to ensure proper application of styles.
- Adjusted CSS selectors to reflect changes in class structure for button styles.

* Enhance class handling for checkbox-multiple and radio fields

- Updated Contact_Form_Field to apply block style classes to wrapper divs for checkbox-multiple and radio types.
- Introduced a new method to extract block style classes from the class attribute.
- Modified tests to validate the addition of block style classes in the rendering process.

* Refactor block style registration for checkbox-multiple and radio fields

- Consolidated block style registration for 'jetpack/field-checkbox-multiple' and 'jetpack/field-radio' into a single call.
- Updated comments to clarify the purpose of style registration and future improvements.
- Adjusted class handling in Contact_Form_Field to include new class naming conventions.

* Refactor block style application for checkbox-multiple and radio fields

- Streamlined the handling of block style classes for 'checkbox-multiple' and 'radio' types in Contact_Form_Field.
- Improved code readability by consolidating conditional checks for block style class assignment.
- Ensured consistent application of styles to wrapper divs for both field types.

* Enhance radio field rendering tests and block style application

- Added tests to validate rendering of radio fields with block style classes.
- Improved handling of block style classes for radio fields in Contact_Form_Field to ensure consistent application.
- Updated assertions in tests to reflect changes in class structure for radio fields.

* Remove outdated changelog entry for migrating form fields to inner blocks with multiple wrapper styles

* Enhance field wrapper class handling for contact form fields

- Introduced 'fieldwrapperclasses' attribute to allow custom wrapper classes for form fields.
- Updated rendering logic in Contact_Form_Field to utilize the new attribute for consistent styling.
- Refactored tests to validate the addition of 'fieldwrapperclasses' for various field types, ensuring proper application of block styles.

* Try fixing a linting issue with the DataProvider

---------

Co-authored-by: Andrew Serong <14988353+andrewserong@users.noreply.github.com>

andrewserong committed:
Add icons for input, label, option, and options blocks (#43270)

talldan committed:
Form Global Styles: Update CSS specificity to 0-1-0 for default editor styles (excluding outlined style) (#43125)

* Reduce specificity of text area styles to 0-1-0

* Update specificity of option block styles

* Fix option input not scaling with font size for single/multiple choice fields

* Fix font:inherit rule overriding global styles font family setting by reduce specificity

* Add dead code comments

ramonjd committed:
Forms: only apply `wp-` and `is-style-` classes to fields of new inner block types (#43268)

* Only apply the block style classes to the field wrapper if the field is one of the new inner block types.

* update tests

* Refactor get_block_style_classes to return field wrapper classes and simplify class handling

* Ensure there's no whitespace in unimgrated fields

simison commented:
FYI, I added label to disable code coverage test as things like that can be worked on separate PRs, but feel free to remove label if you need to run the check.

<img width="454" alt="image" src="https://github.com/user-attachments/assets/c4919598-9521-4e5e-8eef-4fe24d509edd" />

talldan committed:
Fix missing checkmark for single choice options on frontend (#43301)

talldan committed:
Forms: Fix CSS specificity for compatibility with global styles on the frontend (#43299)

* Fix consent field label css specificity preventing global typography styles

* Fix checkbox/radio option font-weight css specificity preventing global styles

* Remove empty line and duplicate classname

* Fix mixed spaces/tabs indentation

* Fix checkbox/radio input not scaling with font size for single/multiple options fields

* Fix indentation

* Fix button style multiple choice checkbox not receiving global styles font color due to css specificity

* Fix single choice option not receiving text color due to styles not applying to correct element

* Reinstate missing color style

* Move option block style classnames to parent wrapper on frontend

* Make option label font weight inherited from parent block wrapper, add some back compat for default css styles

* Fix tests

* Move option styles to parent block level element

ramonjd committed:
Forms: remove color gradients block supports (#43427)

* Forms: Update gradient property naming in contact form block and tests to ensure consistency

* Forms: Update gradient property naming to 'gradients' for consistency in input and option blocks

* Fix label typo

---------

Co-authored-by: Andrew Serong <14988353+andrewserong@users.noreply.github.com>

ramonjd committed:
[WIP] Forms: migrate Outlined form style for inner block based fields (#43327)

* first commit

* backend mania

* Removing dividing two top of label. Can reinstate

* fix number conversion

* Match legacy form behaviour when deriving px values for CSS vars

* color, typography

* Adjust clip path to cater for 1px borders

* Implement outline style support for form fields by integrating useFormStyleOutlineClassesAndStyles hook across various blocks. Update CSS variables and styles for outlined and animated form styles in the contact form backend and frontend components.

* Cannot fix overlap right now

* Trying to fix select, options, etc

* desperately trying to assign the right values to css vars

* get background working

* path to syncing options

* use lodash for merging

* global styles specificity

* Remove isSynced from hook and related code

* Try only using hook on label

* Inject CSS vars for input as well as label so that it can calculate correct padding

* Call it a jetpack contact form notch width

* Fix overflowing background when a narrow border-width is set

* Fix input block placeholders not selectable/visible when outlined style is active

* Fix animated style label and input offset

* Animated style: fix label and input offset when using distinct border left style in editor

* Fix trailing notch background color

* Inputs should take border radius styles so they match the notched border. Fix global style derivation for borders

* Fix incorrect block names

Co-authored-by: Ramon <ramonjd@users.noreply.github.com>

* Improve support for split border radii in outlined form style

* Remove min width for notch

* This commit syncs jetpack/options with jetpack/input

* abstract get inner block hook

* esc notched attrs

* Check $legacy values before assignment

* move css vars to inner elems

* Apply the background to the notch elements again instead of the input to ensure border aligns with background. Remove the z-index rule so that the input/placeholder is visible and can be edited

* Adding padding where there's a border

Enhance contact form styles and structure by adding support for animated styles, adjusting border colors, and refining fieldset rendering logic.

* Add filler notch element on frontend

* Fix flex shrink/grow of notch elements with new filler notch element

* Limit padding of inputs to support a wider range of border radii units

This helps prevent situations where the calculated padding can end up
being a huge number when units like % are used. 100% might make sense for
a border radius, but that value used for a padding is huge. Cap the padding
at 100px to match trunk's max border radius value.

* Ensure dropdowns have same styles for input as other fields, DRY up

* Updated rendering logic in the contact form fields to differentiate between outlined and animated styles. Introduced a new helper function to manage outlined style attributes and adjusted CSS to ensure proper styling for outlined fields. Cleaned up related code for clarity and maintainability.

* Ensure default padding for outlined styles in inputs

* Refactor contact form styles to ensure global styles can override default border and padding settings. Updated class logic for option groups to conditionally apply border styles based on attributes, enhancing flexibility and maintainability.

* Account for split border radius when calculating css var on the frontend

* Set a max width for any elements using --jetpack--contact-form--notch-width of 100px

* Update tests with new shortcode attributes and styles for outline and options

* Switch inputs and labels for the outlined and animated style so that label is above input

* Remove background on inputs for outlined style, background is being applied to labels instead

* Add filler notch element to match editor styles

* Add max-width for trailing notch element

* Fix border radius css var not picking up single values with a unit

* Remove border radius and clip path on inputs for outlined style (border radius is applied on the label, clip path does nothing)

* Update tests with new attributes and outline HTML

* Default white background in the editor

* Ensure the Button variation on the Multiple Choice (Checkbox) block inherits the current color from the option block.

* Add background color support for outlined style in contact form inputs and update related tests

* Fix outlined border radius for checkbox multiple

* Fix css border-radius var calculation

* Fix checkbox multiple radius for outlined style

* Small tidy up

* Enhance outline styles handling in contact form fields by adding form style parameter and refining border radius calculations. Update CSS to include z-index for animated labels.

* Apply outlined border radius style to radio fields to match other fields

* Remove spaces in inline styles

* Select arrow position, default padding for has-background

* Padding magic on the options in outlined

* Add font size adjustment for legend in outlined contact form style

* Fix position of legend and right padding for options

* Remove export of constant

* Combine is_outlined `if` statements for checkbox multiple

* Combine is_outlined `if` statements for radio

* Rename useFormStyleOutlineClassesAndStyles to useVariationStyleProperties

* Only memoize based on changes to block type when merging global styles

* Add documentation to useVariationStyleProperties

* Small tidy up, rename a function and add docs

* Remove commented out css declarations

* Remove dead code. Context is unused. Anywhere. Go look if you don't believe me.

---------

Co-authored-by: Daniel Richards <daniel.richards@automattic.com>

andrewserong committed:
Add esc_attr calls to render_radio_field (#43488)

ramonjd committed:
Reinstate default padding where there's a border. Add option chaining to innerBlock look up as trunk forms won't have it.

talldan committed:
Use correct block name in get_ouline_styles (#43489)

ramonjd committed:
Forms: standardize animated variation style paddings (#43469)

* Update padding for animated style

* Fix updated function

* Rename get_outline_styles

* check borderradius

* Calculate border widths with all sides via CSS vars

* Fix text area top position

* Reinstate default (trunk) padding for animated labels etc

* Default value of null on border sizes

* Remove dupe CSS

* JS Doc :success:

* Add legacy CSS var as default

* Use the `borderwidth` field attribute as the legacy legacy border width value

* Explain use of `borderwidth` value

* Update projects/packages/forms/src/contact-form/class-contact-form-field.php

Use var `$border_width_attribute`

* Add bottom width CSS var to backend, and use border sides in the notched CSS

* Remove styles that we're overriding

* Switch from !empty to isset for legacy borderwidth attribute check

* Do empty/zero check on legacy border widths

---------

Co-authored-by: Daniel Richards <daniel.richards@automattic.com>

talldan committed:
Forms: Fix block instance border color not overriding global styles border color with outlined style active (on the frontend) (#43532)

* Allow block instance border color to override global styles border color

* Remove extra space in class name

ramonjd committed:
Revert input + option syncing. Bad!

talldan committed:
Forms: Fix typography input styles applying to labels when outline style is active (on frontend) (#43534)

* Fix rogue typography attributes from input applying to label

* Fix bad rebase

* Update tests

ramonjd committed:
For classic themes, do not destructure theme config objects that don't exist.

andrewserong committed:
Forms: Fix custom classnames support in inner blocks migration (#43564)

ramonjd committed:
Forms: fix custom class handling in contact form plugin (#43587)

* Forms: Enhance class handling in contact form plugin and tests

- Updated class handling in `class-contact-form-plugin.php` to filter empty block style classes.
- Modified test cases in `Contact_Form_Plugin_Test.php` and `Contact_Form_Test.php` to include additional attributes and ensure proper class wrapping for various field types.
- Improved assertions to validate class attributes in rendered output.

* Make innerblocks optional. Just for linting

talldan committed:
Forms: Fix outline style label font size (#43590)

* Fix text input-style field label font size when outlined style is active.

* Remove scaling using variable

* Fix scaling of multiple option fields when using outlined style

* Tidy up styles

* Wrap labels

* Fix outlined label font size and animation

* Fix animation by ensuring top is a valid value

* Remove will-change: transform on frontend too. Transform usually does not change.

ramonjd committed:
Forms: fix inconsistent weights in outline style and terms labels (#43545)

* Ensure that existing font-weight styles do not override block and global label styles.

* Consent weights

* Revert options positioning kram

* Increase specificity to 0 1 0 for classic themes

* Bump specificity of consent label

* Bump specificity

talldan committed:
Forms: Fix animated style label font scaling (#43615)

* Fix animated style label font size in editor

* Add span wrapper around label text

* Update animated label styles on frontend

* Size and animation adjustments for required text

* Fix label animating when selected in editor if using the animated form style variation and no placeholder is set

talldan committed:
Forms Fix animated label centered positioning (#43635)

* Fix animated label vertical centering when using uneven borders in the input

* Remove unused cass var

talldan committed:
Forms: Remove options text color support (#43612)

* Remove text color support for jetpack/options, since frontend does not replicate the markup of the editor and so style parity cannot be achieved

* Try to fix tests

* Whoops, wrong flag

* One more time

---------

Co-authored-by: Andrew Serong <14988353+andrewserong@users.noreply.github.com>

ramonjd committed:
Forms: options block layout tweaks (#43531)

* Refactor contact form styles and rendering logic to enhance alignment and maintainability. Update CSS for radio and checkbox options to use consistent alignment, and streamline fieldset and div class attributes by removing unnecessary prefixes. Improve handling of border styles and attributes for outlined forms, ensuring compatibility with global styles.

* Try baseline for the options

* Revert changes to "fix" option alignment

* Revert option alignment v2

* Add classes to tests

* Reinstate tweaks for option positioning

* Apply wp-block-jetpack-option class in the attribute migration method

* Use em instead of rem. Remove unnecessary tweaks

* Border support for classic themes

* Ensure radio inputs don't warp in classic themes. Update options description

* Force opinionated dimensions on the options inputs

* wp-block-jetpack-option was not being applied to inner options blocks

ramonjd committed:
Forms: update padding for jetpack options in contact form to conform with trunk (#43634)

* consolidate legends

* Remove styles that are inconsistent with existing trunk spacing. (frontend)

* Remove padding in editor

* editor

* The feature branch has different markup for .wp-block-jetpack-options so we need to target the fieldset using different selectors

* comments

talldan committed:
Fix date field label suffix font size for outlined/animated styles

talldan committed:
Fix option input color not inherited in editor

ramonjd committed:
Forms: Update border radius and width handling in deprecateFieldStyles function (#43653)

* Attempt to fix warnings about PhanUndeclaredTypeParameter for WP_Block

* Forms: Update border radius and width handling in deprecateFieldStyles function

* Ensure border radius and width are correctly set to 'undefined' only when they are not provided or are null, allowing for proper handling of zero values.

* Coerce lineheight to str

* REvert  - bad rebase

* Update class-contact-form-plugin.php

My IDE auto formatted this

---------

Co-authored-by: Andrew Serong <14988353+andrewserong@users.noreply.github.com>

ramonjd committed:
Remove duplicated styles

ramonjd committed:
Form: fix default fallbacks for unmigrated forms (#43676)

* Outlined multi option backgrounds defaults

* Don't concat the label and option styles for consent and single option in unmigrated forms

* PHPCS warning - Use statements are not alphabetically sorted (MediaWiki.Classes.UnsortedUseStatements.UnsortedUse

ramonjd committed:
Forms: tame select arrow positioning (#43674)

* initial commit - adding extra markup to the select rendering and fixing positioning of labels

* uncommenting

* Remove default select border

* Update select field rendering to include a wrapper div with the correct class structure for improved styling and consistency in tests.

* Update contact form styles to inherit text color and adjust z-index for select wrapper elements

* Only run `getAttribute` on select wrapper if there is one

* Reinstate padding change

* phan me

* phanos

* whitespace

ramonjd committed:
Ensure that animated labels sit above active input elements.

ramonjd committed:
Forms: account for missing placeholder when transforming input fields to select dropdown and fix z-index (#43721)

* Enhance select field transformation by adding default placeholder for input block. Update contact form plugin to correctly reference placeholder attribute. Adjust z-index values in CSS for improved label positioning.

* Update z-index values in contact form CSS for improved label positioning and remove redundant styles for outlined fields.

* Complete CSS comment

* Refactor select field transformation to set placeholder attribute directly from attributes. Remove redundant placeholder handling in inner block creation.

* Revert createTextFieldInnerBlocks

* Revert experiment. Add default placeholder for select in createTextFieldInnerBlocks

* Don't mutate the attributes object

* Don't spread optional chaining just in case

ramonjd committed:
Update contact form styles to prevent color overrides by global styles. Add 'color: unset !important;' to both editor.scss and grunion.css for consistent input appearance. (#43737)

talldan committed:
Fix consent field spacing between checkbox and label (#43743)

talldan committed:
Update context types to use kebab casing (#43740)

andrewserong committed:
Forms: Fix required form field validation for options blocks in the inner blocks refactor (#43738)

talldan committed:
Update accessible form selectors for single/multiple choice fields (#43760)

ramonjd committed:
Forms: refactor placeholder handling for outlined and styled forms (#43759)

* Forms: Refactor placeholder handling in contact form fields

* Simplify placeholder logic to ensure consistent rendering of input fields.
* Remove unnecessary conditionals for placeholder assignment and improve code clarity.

* Forms: Update contact form styles and placeholder handling

* Adjusted placeholder logic to ensure it only renders when a value is present, improving input field consistency.
* Updated the contact form stylesheet to include a version query string for better cache management.
* Commented out specific CSS rules to prevent unintended side effects related to placeholder visibility in outlined and animated label styles.

* Remove other :not(:placeholder-shown) styles

talldan committed:
Remove multiple changelogs for branches that were never merged to trunk
